File: src/main/java/leetcode92/Solution.go
package main

/*
反转从位置 m 到 n 的链表。请使用一趟扫描完成反转。

说明:
1 ≤m≤n≤ 链表长度。

示例:

输入: 1->2->3->4->5->NULL, m = 2, n = 4
输出: 1->4->3->2->5->NULL

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/reverse-linked-list-ii
著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
*/
func main() {

}

/**
 * Definition for singly-linked list.
 */
type ListNode struct {
	Val  int
	Next *ListNode
}

func reverseBetween(head *ListNode, left int, right int) *ListNode {
	dumNode := &ListNode{}
	dumNode.Next = head
	pre := dumNode
	for i := 0; i < left-1; i++ {
		pre = pre.Next
	}

	rightNode := pre
	for i := 0; i < right-left+1; i++ {
		rightNode = rightNode.Next
	}

	leftNode := pre.Next
	cur := rightNode.Next

	pre.Next = nil
	rightNode.Next = nil

	reverseLinkedList := func(head *ListNode) {
		//pre ->cur ->next
		var pre *ListNode
		cur := head
		for cur != nil {
			next := cur.Next
			cur.Next = pre
			pre = cur
			cur = next
		}
	}
	reverseLinkedList(leftNode)

	pre.Next = rightNode
	leftNode.Next = cur

	return dumNode.Next
}
func reverseBetween2(head *ListNode, left int, right int) *ListNode {
	dum := &ListNode{}
	dum.Next = head
	pre := dum
	for i := 0; i < left-1; i++ {
		pre = pre.Next
	}
	cur := pre.Next
	//		--next----
	//		|		|
	//		v		|
	// pre->cur ->next->next.next
	//		|			    ^
	//		-----next-------|
	for i := 0; i < right-left; i++ {
		next := cur.Next
		cur.Next = next.Next
		//滚动，所以不能等于cur 需要每次都等于pre的下一个节点next
		next.Next = pre.Next
		pre.Next = next
	}

	return dum.Next
}

File: src/main/java/leetcode92/Solution.rs
/*
反转从位置 m 到 n 的链表。请使用一趟扫描完成反转。

说明:
1 ≤m≤n≤ 链表长度。

示例:

输入: 1->2->3->4->5->NULL, m = 2, n = 4
输出: 1->4->3->2->5->NULL

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/reverse-linked-list-ii
著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
*/
// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
  pub val: i32,
  pub next: Option<Box<ListNode>>
}

impl ListNode {
  #[inline]
  fn new(val: i32) -> Self {
    ListNode {
      next: None,
      val
    }
  }
}
impl Solution {
    pub fn reverse_between(mut head: Option<Box<ListNode>>, left: i32, right: i32) -> Option<Box<ListNode>> {
        let mut dm = Some(Box::new(ListNode { val: 0, next: head }));
        let mut dm_mut=dm.as_mut();
        let mut i = 0;
        // let mut left_node = None;
        let mut pre = None;
        while let Some(mut left_head) = dm_mut {
            i += 1;
            while i >= left && i <= right {
                let mut cur = left_head.next.take();
                left_head.next= cur.as_mut().unwrap().next.take();
                cur.as_mut().unwrap().next= pre.take();
                if i<right {
                    pre= cur;
                }else {
                    let mut rev = cur.as_mut();
                    while let Some(rev_node) = rev {
                        if rev_node.next.is_none() {
                            rev_node.next= left_head.next.take();
                            break;
                        }
                        rev=rev_node.next.as_mut();
                    }
                    left_head.next= cur;
                    return dm.unwrap().next
                }
                i+=1;
            }
            dm_mut = left_head.next.as_mut();
        }

        dm.unwrap().next
    }

}